File: bridges/dynamic_tool.py
from __future__ import annotations

import re
from typing import TYPE_CHECKING

from tools.base import Tool

if TYPE_CHECKING:
    from .base import Bridge
    from .spec import ToolSpec


# OpenAI tool names must match ^[a-zA-Z0-9_-]{1,64}$. We namespace
# device-contributed tools as "<device>__<tool>" to avoid collisions
# with built-in tools and across devices.
_VALID_NAME = re.compile(r"^[a-zA-Z0-9_-]+$")


def namespaced_tool_name(device: str, tool: str) -> str:
    safe = lambda s: re.sub(r"[^A-Za-z0-9_-]", "_", s)
    full = f"{safe(device)}__{safe(tool)}"
    return full[:64]


class BridgeTool(Tool):
    """Adapter — turns a (Bridge, ToolSpec) pair into the in-tree Tool
    interface. The agent sees this exactly like any built-in tool."""

    def __init__(self, bridge: "Bridge", spec: "ToolSpec"):
        self._bridge = bridge
        self._spec = spec
        self.name = namespaced_tool_name(bridge.name, spec.name)
        # Keep the description short but identifiable so the model can
        # tell two devices' "send" tools apart.
        self.description = f"[{bridge.name}] {spec.description}".strip()
        self.parameters = spec.parameters or {
            "type": "object", "properties": {}, "required": []
        }

    async def run(self, **kwargs) -> str:
        try:
            return await self._bridge.invoke(self._spec, kwargs)
        except Exception as e:
            return f"Bridge '{self._bridge.name}' error: {e}"

    @property
    def device_name(self) -> str:
        return self._bridge.name

    @property
    def underlying(self) -> str:
        return self._spec.name

File: cli-ink/src/bridges/dynamic-tool.ts
import type { Tool } from "../tools/tool.js";
import type { Bridge } from "./base.js";
import type { ToolSpec } from "./types.js";

/**
 * Convert non-alphanumeric/_/- characters to `_`, then truncate to 64 —
 * matching OpenAI's tool-name rules. Namespace as `<device>__<tool>`.
 */
export function namespacedToolName(device: string, tool: string): string {
  const safe = (s: string) => s.replace(/[^A-Za-z0-9_-]/g, "_");
  const full = `${safe(device)}__${safe(tool)}`;
  return full.slice(0, 64);
}

/**
 * Adapter: turns `(Bridge, ToolSpec)` into the existing `Tool` interface so
 * the agent calls bridge tools just like any built-in.
 */
export class BridgeTool implements Tool {
  name: string;
  description: string;
  parameters: any;
  private bridge: Bridge;
  private spec: ToolSpec;

  constructor(bridge: Bridge, spec: ToolSpec) {
    this.bridge = bridge;
    this.spec = spec;
    this.name = namespacedToolName(bridge.name, spec.name);
    this.description = `[${bridge.name}] ${spec.description}`.trim();
    this.parameters = spec.parameters && typeof spec.parameters === "object"
      ? spec.parameters
      : { type: "object", properties: {}, required: [] };
  }

  async run(args: Record<string, any>): Promise<string> {
    try {
      const out = await this.bridge.invoke(this.spec, args);
      // Final safety net — if a bridge that doesn't already chunk its
      // output (e.g., http) returns a giant blob, truncate before it
      // hits the model's context. MCPBridge handles this internally with
      // file offloading; this is for everything else.
      const HARD_CAP = 32_000;
      if (out.length > HARD_CAP) {
        return (
          out.slice(0, HARD_CAP - 200) +
          `\n... (truncated — ${out.length - HARD_CAP} more chars dropped to protect context)`
        );
      }
      return out;
    } catch (e: any) {
      return `Bridge '${this.bridge.name}' error: ${e?.message ?? e}`;
    }
  }
}
